keps/sig-node/3953-dynamic-node-resize/README.md

```diff
@@ -0,0 +1,776 @@
+     |                                |                               |                                |
+     |                                |                               |             update             |
+     |                                |                               |<-------------------------------|
+     |                                |                               |                                |
+     |                                |            update             |                                |
+     |                                |<------------------------------|                                |
+     |                                |                               |                                |
+     |                                |                               |                                |
+     |                                |                               |                                |
+     |     node status update         |                               |                                |
+     |<-------------------------------|                               |                                |
+     |                                |                               |                                |
+     |      if shrink in resource     |                               |                                |
+     |      re-run pod admission      |                               |                                |
+     |<-------------------------------|                               |                                |                                |
+     |                                |                               |                                |      
+     | re-initialize resource managers|                               |                                |
+     |<-------------------------------|                               |                                |                                |
+     |                                |                               |                                |           
+
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling cAdvisor in interval of configured time to fetch the machine resource information.
+2. cAdvisor will fetch and update the machine resource information.
+3. Kubelet's cache will be updated with the latest machine resource information.
+4. Node status updater will update the node's status with the latest resource information.
+5. In case of a shrink in cluster resources rerun the pod admission and the pod admission will evict pods
+6. Kubelet will reinitialize the resource managers to keep them up to date with dynamic resource changes.
+
+Note: In case of increase in cluster resources, the scheduler will automatically schedule any pending pods.
+
+**Kubelet Configuration changes**
+
+A new boolean variable will be added to kubelet configuration named "DynamicNodeResize" and will be false by default.
```

Comment:
Is DynamicNodeResize a feature gate of kubelet or a configuration in kubelet configuration?

For a feature gate, we will make it an option to promote it from alpha to beta and then GA as the default behavior. It can not be turned off once it is GAed most likely.
For a flag, we will keep it configurable.

Reply:
I was under assumption to have this feature gate mandatory for an enhancement.
For this case, I believe keeping DynamicNodeResize as a kubelet configuration is more suitable as user can opt in for this feature, only if required.

Reply:
If so, you may follow https://github.com/kubernetes/enhancements/blob/0f16136bac9035e1ad9cd30d4e1d273aa062101e/keps/sig-node/3673-kubelet-parallel-image-pull-limit/README.md#proposal to not use a feature gate.

Reply:
But the question: why not make is default behaviour post GA? this doesn't look like any new feature to me as usage perspective.

Reply:
Initially started with idea of having this enhancement behind feature gate and kubelet config variable,  but now seems like both of these are redundant, based on the community opinion and recommendation I will change it accordingly.

Reply:
@Karthik-K-N perhaps is implicit in the conversation but the feature gate is required by the feature graduation process until the feature reaches GA (which is what @pacoxu mentioned initially)